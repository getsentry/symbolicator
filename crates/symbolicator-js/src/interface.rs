use std::collections::HashSet;
use std::fmt;
use std::sync::Arc;

use reqwest::Url;
use serde::{Deserialize, Serialize};

use symbolicator_service::caching::CacheError;
use symbolicator_service::services::ScrapingConfig;
use symbolicator_service::types::{RawObjectInfo, Scope};
use symbolicator_sources::{SentryFileId, SentrySourceConfig};

#[derive(Debug, Clone)]
pub struct SymbolicateJsStacktraces {
    pub scope: Scope,
    pub source: Arc<SentrySourceConfig>,
    pub release: Option<String>,
    pub dist: Option<String>,
    pub debug_id_index: Option<Url>,
    pub url_index: Option<Url>,
    pub stacktraces: Vec<JsStacktrace>,
    pub modules: Vec<RawObjectInfo>,
    pub scraping: ScrapingConfig,
    /// Whether to apply source context for the stack frames.
    pub apply_source_context: bool,
}

// Some of the renames are there only to make it synchronized
// with the already existing monolith naming scheme.
#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "snake_case")]
#[serde(tag = "type")]
pub enum JsModuleErrorKind {
    InvalidLocation { line: u32, col: Option<u32> },
    InvalidAbsPath,
    NoColumn,
    MissingSourceContent { source: String, sourcemap: String },
    MissingSource,
    MalformedSourcemap { url: String },
    MissingSourcemap,
    InvalidBase64Sourcemap,
    ScrapingDisabled,
}

impl fmt::Display for JsModuleErrorKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            JsModuleErrorKind::InvalidLocation { line, col } => {
                write!(f, "Invalid source location")?;
                match (line, col) {
                    (l, None) => write!(f, ": line:{l}")?,
                    (l, Some(c)) => write!(f, ": line:{l}, col:{c}")?,
                }
                Ok(())
            }
            JsModuleErrorKind::InvalidAbsPath => write!(f, "Invalid absolute path"),
            JsModuleErrorKind::NoColumn => write!(f, "No column information"),
            JsModuleErrorKind::MissingSourceContent { source, sourcemap } => write!(
                f,
                "Missing source contents for source file {source} and sourcemap file {sourcemap}"
            ),
            JsModuleErrorKind::MissingSource => write!(f, "Missing source file"),
            JsModuleErrorKind::MalformedSourcemap { url } => {
                write!(f, "Sourcemap file at {url} is malformed")
            }
            JsModuleErrorKind::MissingSourcemap => write!(f, "Missing sourcemap file"),
            JsModuleErrorKind::InvalidBase64Sourcemap => write!(f, "Invalid base64 sourcemap"),
            JsModuleErrorKind::ScrapingDisabled => {
                write!(f, "Could not download file because scraping is disabled")
            }
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq, PartialOrd, Ord)]
pub struct JsModuleError {
    pub abs_path: String,
    #[serde(flatten)]
    pub kind: JsModuleErrorKind,
}

/// An attempt to scrape a JS source or sourcemap file from the web.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct JsScrapingAttempt {
    /// The URL we attempted to scrape from.
    pub url: String,
    /// The outcome of the attempt.
    #[serde(flatten)]
    pub result: JsScrapingResult,
}

impl JsScrapingAttempt {
    pub fn success(url: String) -> Self {
        Self {
            url,
            result: JsScrapingResult::Success,
        }
    }
    pub fn not_attempted(url: String) -> Self {
        Self {
            url,
            result: JsScrapingResult::NotAttempted,
        }
    }

    pub fn failure(url: String, reason: JsScrapingFailureReason, details: String) -> Self {
        Self {
            url,
            result: JsScrapingResult::Failure { reason, details },
        }
    }
}

/// The outcome of a scraping attempt.
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "snake_case")]
#[serde(tag = "status")]
pub enum JsScrapingResult {
    /// We didn't actually attempt scraping because we already obtained the file
    /// by another method.
    NotAttempted,
    /// The file was succesfully scraped.
    Success,
    /// The file couldn't be scraped.
    Failure {
        /// The basic reason for the failure.
        reason: JsScrapingFailureReason,
        #[serde(skip_serializing_if = "String::is_empty")]
        /// A more detailed explanation of the failure.
        details: String,
    },
}

impl From<CacheError> for JsScrapingResult {
    fn from(value: CacheError) -> Self {
        let (reason, details) = match value {
            CacheError::NotFound => (JsScrapingFailureReason::NotFound, String::new()),
            CacheError::PermissionDenied(details) => {
                (JsScrapingFailureReason::PermissionDenied, details)
            }
            CacheError::Timeout(duration) => (
                JsScrapingFailureReason::Timeout,
                format!("Timeout after {}", humantime::format_duration(duration)),
            ),
            CacheError::DownloadError(details) => (JsScrapingFailureReason::DownloadError, details),
            CacheError::Malformed(details) => (JsScrapingFailureReason::Other, details),
            CacheError::InternalError => (JsScrapingFailureReason::Other, String::new()),
        };

        Self::Failure { reason, details }
    }
}

/// The basic reason a scraping attempt failed.
#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum JsScrapingFailureReason {
    /// The file was not found at the given URL.
    NotFound,
    /// Scraping was disabled.
    Disabled,
    /// The URL was not in the list of allowed hosts or had
    /// an invalid scheme.
    InvalidHost,
    /// Permission to access the file was denied.
    PermissionDenied,
    /// The scraping attempt timed out.
    Timeout,
    /// There was a non-timeout error while downloading.
    DownloadError,
    /// Catchall case.
    ///
    /// This probably can't actually happen.
    Other,
}

#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct JsFrame {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub function: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub filename: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub module: Option<String>,

    pub abs_path: String,

    pub lineno: u32,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub colno: Option<u32>,

    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub pre_context: Vec<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub context_line: Option<String>,

    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub post_context: Vec<String>,

    #[serde(skip_serializing)]
    pub token_name: Option<String>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub in_app: Option<bool>,

    #[serde(default, skip_serializing_if = "JsFrameData::is_empty")]
    pub data: JsFrameData,
}

#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct JsFrameData {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sourcemap: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resolved_with: Option<ResolvedWith>,
    #[serde(default)]
    pub symbolicated: bool,
}

/// A marker indicating what a File was resolved with.
///
/// This enum serves a double purpose, both marking how an individual file was found inside of a
/// bundle, as well as tracking through which method that bundle itself was found.
///
#[derive(Debug, Default, Clone, Copy, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub enum ResolvedWith {
    /// Both: Found in a Bundle via DebugId
    /// And: Found the Bundle via API Lookup via DebugId / Database Index
    DebugId,
    /// Found in a Bundle via Url matching
    Url,
    /// Found the Bundle via API Lookup via Database Index
    Index,
    /// Found the File in a Flat File / Bundle Index
    BundleIndex,
    /// Found the Bundle via API Lookup as an ArtifactBundle
    Release,
    /// Found the Bundle via API Lookup as a ReleaseFile
    ReleaseOld,
    /// Scraped the File from the Web
    Scraping,
    /// Unknown
    #[default]
    Unknown,
}

impl JsFrameData {
    pub fn is_empty(&self) -> bool {
        *self == Self::default()
    }
}

#[derive(Clone, Debug, Default, Deserialize, Serialize)]
pub struct JsStacktrace {
    pub frames: Vec<JsFrame>,
}

#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct CompletedJsSymbolicationResponse {
    pub stacktraces: Vec<JsStacktrace>,
    pub raw_stacktraces: Vec<JsStacktrace>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub errors: Vec<JsModuleError>,
    #[serde(skip_serializing_if = "HashSet::is_empty")]
    pub used_artifact_bundles: HashSet<SentryFileId>,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub scraping_attempts: Vec<JsScrapingAttempt>,
}
