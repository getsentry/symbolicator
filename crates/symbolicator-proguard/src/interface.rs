use std::sync::Arc;

use serde::{Deserialize, Serialize};
use symbolic::common::DebugId;
use symbolicator_service::types::Scope;
use symbolicator_sources::SourceConfig;

/// A request for symbolication/remapping of a JVM event.
#[derive(Debug, Clone)]
pub struct SymbolicateJvmStacktraces {
    /// The scope of this request which determines access to cached files.
    pub scope: Scope,
    /// A list of external sources to load debug files.
    pub sources: Arc<[SourceConfig]>,
    /// The exceptions to symbolicate/remap.
    pub exceptions: Vec<JvmException>,
    /// The list of stacktraces to symbolicate/remap.
    pub stacktraces: Vec<JvmStacktrace>,
    /// A list of proguard files to use for remapping.
    pub modules: Vec<JvmModule>,
    /// Whether to apply source context for the stack frames.
    pub apply_source_context: bool,
    /// The package name of the release this event belongs to.
    ///
    /// This is used to set a remapped frame's `in_app` field.
    pub release_package: Option<String>,
}

/// A stack frame in a JVM stacktrace.
#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct JvmFrame {
    /// The frame's function name.
    ///
    /// For a JVM frame, this is always a class method.
    pub function: String,

    /// The source file name.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub filename: Option<String>,

    /// The frame's method name.
    ///
    /// For a JVM frame, this is a fully qualified class name.
    pub module: String,

    /// The source file's absolute path.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub abs_path: Option<String>,

    /// The line number within the source file, starting at `1` for the first line.
    pub lineno: u32,

    /// Source context before the context line.
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub pre_context: Vec<String>,

    /// The context line if available.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context_line: Option<String>,

    /// Post context after the context line
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub post_context: Vec<String>,

    /// Whether the frame is related to app-code (rather than libraries/dependencies).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub in_app: Option<bool>,
}

/// An exception in a JVM event.
#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct JvmException {
    /// The type (class name) of the exception.
    #[serde(rename = "type")]
    pub ty: String,
    /// The module in which the exception is defined.
    pub module: String,
}

/// A JVM stacktrace.
#[derive(Debug, Default, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct JvmStacktrace {
    /// The stacktrace's frames.
    pub frames: Vec<JvmFrame>,
}

/// A JVM module (proguard file).
#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct JvmModule {
    /// The file's UUID.
    ///
    /// This is used to download the file from symbol sources.
    pub uuid: DebugId,
}

// TODO: Expand this
/// The symbolicated/remapped event data.
#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
pub struct CompletedJvmSymbolicationResponse {
    /// The exceptions after remapping.
    pub exceptions: Vec<JvmException>,
    /// The stacktraces after remapping.
    pub stacktraces: Vec<JvmStacktrace>,
    /// Errors that occurred during symbolication.
    pub errors: Vec<(DebugId, String)>,
}
