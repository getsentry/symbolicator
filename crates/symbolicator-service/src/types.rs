//! Types for the Symbolicator API.
//!
//! This module contains some types which (de)serialise to/from JSON to make up the public
//! HTTP API.  Its messy and things probably need a better place and different way to signal
//! they are part of the public API.

use std::collections::BTreeMap;
use std::fmt;
use std::sync::Arc;

use serde::{Deserialize, Serialize};
use symbolicator_sources::ObjectType;

use crate::utils::hex::HexValue;

/// The scope of a source or debug file.
///
/// Based on scopes, access to debug files that have been cached is determined. If a file comes from
/// a public source, it can be used for any symbolication request. Otherwise, the symbolication
/// request must match the scope of a file.
#[derive(Debug, Clone, Deserialize, Serialize, Eq, Ord, PartialEq, PartialOrd, Hash)]
#[serde(untagged)]
#[derive(Default)]
pub enum Scope {
    #[serde(rename = "global")]
    #[default]
    Global,
    Scoped(Arc<str>),
}

impl AsRef<str> for Scope {
    fn as_ref(&self) -> &str {
        match *self {
            Scope::Global => "global",
            Scope::Scoped(ref s) => s,
        }
    }
}

impl fmt::Display for Scope {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match *self {
            Scope::Global => f.write_str("global"),
            Scope::Scoped(ref scope) => f.write_str(scope),
        }
    }
}

/// Configuration for scraping of JS Sources, Source Maps and Source Context from the web.
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ScrapingConfig {
    /// Whether scraping should happen at all.
    pub enabled: bool,
    /// Whether Symbolicator should verify SSL certs when scraping from the web.
    ///
    /// Defaults to `true`, just to be safe.
    #[serde(default = "default_verify_ssl")]
    pub verify_ssl: bool,
    /// A list of "allowed origin patterns" that control:
    /// - for sourcemaps: what URLs we are allowed to scrape from.
    /// - for source context: which URLs should be authenticated using attached headers
    ///
    /// Allowed origins may be defined in several ways:
    /// - `http://domain.com[:port]`: Exact match for base URI (must include port).
    /// - `*`: Allow any domain.
    /// - `*.domain.com`: Matches domain.com and all subdomains, on any port.
    /// - `domain.com`: Matches domain.com on any port.
    /// - `*:port`: Wildcard on hostname, but explicit match on port.
    pub allowed_origins: Vec<String>,
    /// A map of headers to send with every HTTP request while scraping.
    pub headers: BTreeMap<String, String>,
}

impl Default for ScrapingConfig {
    fn default() -> Self {
        Self {
            enabled: true,
            // verify_ssl: false,
            allowed_origins: vec!["*".to_string()],
            headers: Default::default(),
            verify_ssl: true,
        }
    }
}

fn default_verify_ssl() -> bool {
    true
}

/// Specification of a module loaded into the process.
#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
pub struct RawObjectInfo {
    /// Platform image file type (container format).
    #[serde(rename = "type")]
    pub ty: ObjectType,

    /// Identifier of the code file.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub code_id: Option<String>,

    /// Name of the code file.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub code_file: Option<String>,

    /// Identifier of the debug file.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub debug_id: Option<String>,

    /// Name of the debug file.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub debug_file: Option<String>,

    /// Checksum of the file's contents.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub debug_checksum: Option<String>,

    /// Absolute address at which the image was mounted into virtual memory.
    ///
    /// We do allow the `image_addr` to be skipped if it is zero. This is because systems like WASM
    /// do not require modules to be mounted at a specific absolute address. Per definition, a
    /// module mounted at `0` does not support absolute addressing.
    #[serde(default)]
    pub image_addr: HexValue,

    /// Size of the image in virtual memory.
    ///
    /// The size is infered from the module list if not specified.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub image_size: Option<u64>,
}

/// Information on a debug information file.
#[derive(Debug, Clone, Copy, Serialize, Deserialize, Eq, PartialEq)]
#[serde(rename_all = "snake_case")]
#[derive(Default)]
pub enum ObjectFileStatus {
    /// The file was found and successfully processed.
    Found,
    /// The image was not referenced in the stack trace and not further handled.
    #[default]
    Unused,
    /// The file could not be found in any of the specified sources.
    Missing,
    /// The file failed to process.
    Malformed,
    /// The file could not be downloaded.
    FetchingFailed,
    /// Downloading or processing the file took too long.
    Timeout,
    /// An internal error while handling this image.
    Other,
}

impl ObjectFileStatus {
    pub fn name(self) -> &'static str {
        // used for metrics
        match self {
            ObjectFileStatus::Found => "found",
            ObjectFileStatus::Unused => "unused",
            ObjectFileStatus::Missing => "missing",
            ObjectFileStatus::Malformed => "malformed",
            ObjectFileStatus::FetchingFailed => "fetching_failed",
            ObjectFileStatus::Timeout => "timeout",
            ObjectFileStatus::Other => "other",
        }
    }
}
