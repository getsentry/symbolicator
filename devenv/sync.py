from devenv import constants
from devenv.lib import config, proc, uv


def main(context: dict[str, str]) -> int:
    reporoot = context["reporoot"]
    cfg = config.get_repo(reporoot)

    uv.install(
        cfg["uv"]["version"],
        cfg["uv"][constants.SYSTEM_MACHINE],
        cfg["uv"][f"{constants.SYSTEM_MACHINE}_sha256"],
        reporoot,
    )

    print("syncing .venv ...")
    proc.run(
        (f"{reporoot}/.devenv/bin/uv", "sync", "--frozen", "--quiet", "--active"),
    )

    print("installing pre-commit hooks ...")
    proc.run((f"{reporoot}/.venv/bin/pre-commit", "install", "--install-hooks"))

    return 0
