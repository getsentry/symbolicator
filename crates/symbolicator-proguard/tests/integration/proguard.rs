use std::sync::Arc;
use std::{collections::HashMap, str::FromStr};

use serde_json::json;
use symbolic::common::{DebugId, Uuid};
use symbolicator_proguard::interface::{
    JvmFrame, JvmModule, JvmStacktrace, SymbolicateJvmStacktraces,
};
use symbolicator_service::types::Scope;
use symbolicator_sources::{SentrySourceConfig, SourceConfig};
use symbolicator_test::assert_snapshot;

use crate::setup_service;

fn make_jvm_request(
    source: SourceConfig,
    exception: &str,
    frames: &str,
    modules: &str,
    release_package: impl Into<Option<String>>,
) -> SymbolicateJvmStacktraces {
    let exceptions = vec![serde_json::from_str(exception).unwrap()];
    let frames: Vec<JvmFrame> = serde_json::from_str(frames).unwrap();
    let modules: Vec<JvmModule> = serde_json::from_str(modules).unwrap();
    let stacktraces = vec![JvmStacktrace { frames }];

    SymbolicateJvmStacktraces {
        scope: Scope::Global,
        sources: Arc::new([source]),
        apply_source_context: true,
        release_package: release_package.into(),

        exceptions,
        stacktraces,
        modules,
    }
}

fn proguard_server<L>(
    fixtures_dir: &str,
    lookup: L,
) -> (symbolicator_test::Server, SentrySourceConfig)
where
    L: Fn(&str, &HashMap<String, String>) -> serde_json::Value + Clone + Send + 'static,
{
    let fixtures_dir = symbolicator_test::fixture(format!("proguard/{fixtures_dir}"));
    symbolicator_test::sentry_server(fixtures_dir, lookup)
}

#[tokio::test]
async fn test_download_proguard_file() {
    symbolicator_test::setup();
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = proguard_server("download_proguard_file", |_url, _query| {
        json!([{
            "id":"proguard.txt",
            "uuid":"246fb328-fc4e-406a-87ff-fc35f6149d8f",
            "debugId":"246fb328-fc4e-406a-87ff-fc35f6149d8f",
            "codeId":null,
            "cpuName":"any",
            "objectName":"proguard-mapping",
            "symbolType":"proguard",
            "headers": {
                "Content-Type":"text/x-proguard+plain"
            },
            "size":3619,
            "sha1":"deba83e73fd18210a830db372a0e0a2f2293a989",
            "dateCreated":"2024-02-14T10:49:38.770116Z",
            "data":{
                "features":["mapping"]
            }
        }])
    });

    let source = SourceConfig::Sentry(Arc::new(source));
    let debug_id = DebugId::from_str("246fb328-fc4e-406a-87ff-fc35f6149d8f").unwrap();

    assert!(symbolication
        .download_proguard_file(&[source], &Scope::Global, debug_id)
        .await
        .is_ok());
}

#[tokio::test]
async fn test_remap_exception() {
    symbolicator_test::setup();
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = proguard_server("remap_exception", |_url, _query| {
        json!([{
            "id":"proguard.txt",
            "uuid":"246fb328-fc4e-406a-87ff-fc35f6149d8f",
            "debugId":"246fb328-fc4e-406a-87ff-fc35f6149d8f",
            "codeId":null,
            "cpuName":"any",
            "objectName":"proguard-mapping",
            "symbolType":"proguard",
            "headers": {
                "Content-Type":"text/x-proguard+plain"
            },
            "size":3619,
            "sha1":"deba83e73fd18210a830db372a0e0a2f2293a989",
            "dateCreated":"2024-02-14T10:49:38.770116Z",
            "data":{
                "features":["mapping"]
            }
        }])
    });

    let source = SourceConfig::Sentry(Arc::new(source));

    let request = make_jvm_request(
        source,
        r#"{"type": "g$a", "module": "org.a.b"}"#,
        "[]",
        r#"[{"uuid": "246fb328-fc4e-406a-87ff-fc35f6149d8f", "type": "proguard"}]"#,
        None,
    );

    let response = symbolication.symbolicate_jvm(request).await;

    assert_snapshot!(response);
}

#[tokio::test]
async fn test_resolving_inline() {
    symbolicator_test::setup();
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = proguard_server("resolving_inline", |_url, _query| {
        json!([{
            "id":"proguard.txt",
            "uuid":"246fb328-fc4e-406a-87ff-fc35f6149d8f",
            "debugId":"246fb328-fc4e-406a-87ff-fc35f6149d8f",
            "codeId":null,
            "cpuName":"any",
            "objectName":"proguard-mapping",
            "symbolType":"proguard",
            "headers": {
                "Content-Type":"text/x-proguard+plain"
            },
            "size":3619,
            "sha1":"deba83e73fd18210a830db372a0e0a2f2293a989",
            "dateCreated":"2024-02-14T10:49:38.770116Z",
            "data":{
                "features":["mapping"]
            }
        }])
    });

    let source = SourceConfig::Sentry(Arc::new(source));
    let frames = r#"[{
        "function": "onClick",
        "module": "e.a.c.a",
        "lineno": 2,
        "index": 0
    }, {
        "function": "t",
        "module": "io.sentry.sample.MainActivity",
        "filename": "MainActivity.java",
        "lineno": 1,
        "index": 1
    }]"#;

    let request = make_jvm_request(
        source,
        r#"{"type": "g$a", "module": "org.a.b"}"#,
        frames,
        r#"[{"uuid": "246fb328-fc4e-406a-87ff-fc35f6149d8f", "type": "proguard"}]"#,
        None,
    );

    let response = symbolication.symbolicate_jvm(request).await;

    assert_snapshot!(response);
}

#[tokio::test]
async fn test_basic_source_lookup() {
    symbolicator_test::setup();
    let (symbolication, _cache_dir) = setup_service(|_| ());
    let (_srv, source) = proguard_server("basic_source_lookup", |_url, _query| {
        json!([{
            "id":"bundle.zip",
            "uuid":"246fb328-fc4e-406a-87ff-fc35f6149d8f",
            "debugId":"246fb328-fc4e-406a-87ff-fc35f6149d8f",
            "codeId":null,
            "cpuName":"any",
            "objectName":"proguard-mapping",
            "symbolType":"sourcebundle",
            "headers": {
                "Content-Type":"application/x-sentry-bundle+zip"
            },
            "size":3619,
            "sha1":"deba83e73fd18210a830db372a0e0a2f2293a989",
            "dateCreated":"2024-02-14T10:49:38.770116Z",
        }])
    });

    let source = SourceConfig::Sentry(Arc::new(source));

    let frames = r#"[{
        "function": "otherMethod",
        "abs_path": "OtherActivity.java",
        "module": "OtherActivity",
        "filename": "OtherActivity.java",
        "lineno": 100,
        "index": 0
    }, {
        "function": "differentMethod",
        "abs_path": "DifferentActivity",
        "module": "DifferentActivity",
        "filename": "DifferentActivity",
        "lineno": 200,
        "index": 1
    }, {
        "function": "onCreate",
        "module": "io.sentry.samples.MainActivity",
        "lineno": 11,
        "index": 2
    }, {
        "function": "whoops",
        "abs_path": "MainActivity.kt",
        "module": "io.sentry.samples.MainActivity$InnerClass",
        "filename": "MainActivity.kt",
        "lineno": 20,
        "index": 3
    }, {
        "function": "whoops2",
        "module": "io.sentry.samples.MainActivity$AnotherInnerClass",
        "lineno": 26,
        "index": 4
    }, {
        "function": "whoops3",
        "abs_path": "MainActivity.kt",
        "module": "io.sentry.samples.MainActivity$AdditionalInnerClass",
        "filename": "MainActivity.kt",
        "lineno": 32,
        "index": 5
    }, {
        "function": "whoops4",
        "abs_path": "SourceFile",
        "module": "io.sentry.samples.MainActivity$OneMoreInnerClass",
        "filename": "SourceFile",
        "lineno": 38,
        "index": 6
    }]"#;

    let request = make_jvm_request(
        source,
        r#"{"type": "RuntimeException", "module": "io.sentry.samples"}"#,
        frames,
        r#"[{"uuid": "246fb328-fc4e-406a-87ff-fc35f6149d8f", "type": "source"}]"#,
        None,
    );

    let response = symbolication.symbolicate_jvm(request).await;

    assert_snapshot!(response);
}

#[tokio::test]
async fn test_source_lookup_with_proguard() {
    symbolicator_test::setup();
    let (symbolication, _cache_dir) = setup_service(|_| ());

    let proguard_id = "05d96b1c-1786-477c-8615-d3cf83e027c7".parse().unwrap();
    let missing_proguard_id: DebugId = "8236f5cf-52c8-4e35-a7cf-01421e4c2c88".parse().unwrap();
    let source_id1 = DebugId::from_uuid(Uuid::new_v4());
    let source_id2 = DebugId::from_uuid(Uuid::new_v4());

    let (_srv, source) = proguard_server("source_lookup_with_proguard", move |_url, query| {
        let debug_id: DebugId = query["debug_id"].parse().unwrap();

        match debug_id {
            _ if debug_id == source_id1 => {
                json!([{
                    "id":"edit_activity.zip",
                    "debugId":source_id1,
                    "codeId":null,
                    "cpuName":"any",
                    "objectName":"proguard-mapping",
                    "symbolType":"sourcebundle",
                    "headers": {
                        "Content-Type":"application/x-sentry-bundle+zip"
                    },
                    "size":3619,
                    "sha1":"deba83e73fd18210a830db372a0e0a2f2293a989",
                    "dateCreated":"2024-02-14T10:49:38.770116Z",
                }])
            }
            _ if debug_id == source_id2 => {
                json!([{
                    "id":"some_service.zip",
                    "debugId":source_id2,
                    "codeId":null,
                    "cpuName":"any",
                    "objectName":"proguard-mapping",
                    "symbolType":"sourcebundle",
                    "headers": {
                        "Content-Type":"application/x-sentry-bundle+zip"
                    },
                    "size":3619,
                    "sha1":"deba83e73fd18210a830db372a0e0a2f2293a989",
                    "dateCreated":"2024-02-14T10:49:38.770116Z",
                }])
            }
            _ if debug_id == proguard_id => {
                json!([{
                    "id":"proguard.txt",
                    "uuid":proguard_id,
                    "debugId":proguard_id,
                    "codeId":null,
                    "cpuName":"any",
                    "objectName":"proguard-mapping",
                    "symbolType":"proguard",
                    "headers": {
                        "Content-Type":"text/x-proguard+plain"
                    },
                    "size":3619,
                    "sha1":"deba83e73fd18210a830db372a0e0a2f2293a989",
                    "dateCreated":"2024-02-14T10:49:38.770116Z",
                    "data":{
                        "features":["mapping"]
                    }
                }])
            }
            _ => json!([]),
        }
    });

    let source = SourceConfig::Sentry(Arc::new(source));

    let frames = r#"[{
        "filename": "ZygoteInit.java",
        "function": "main",
        "module": "com.android.internal.os.ZygoteInit",
        "lineno": 698,
        "index": 0
    }, {
        "filename": "ZygoteInit.java",
        "function": "run",
        "module": "com.android.internal.os.ZygoteInit$MethodAndArgsCaller",
        "lineno": 903,
        "index": 1
    }, {
        "filename": "Method.java",
        "function": "invoke",
        "module": "java.lang.reflect.Method",
        "lineno": 372,
        "index": 2
    }, {
        "filename": "Method.java",
        "function": "invoke",
        "module": "java.lang.reflect.Method",
        "lineno": 0,
        "index": 3
    }, {
        "filename": "ActivityThread.java",
        "function": "main",
        "module": "android.app.ActivityThread",
        "lineno": 5254,
        "index": 4
    }, {
        "filename": "Looper.java",
        "function": "loop",
        "module": "android.os.Looper",
        "lineno": 135,
        "index": 5
    }, {
        "filename": "Handler.java",
        "function": "dispatchMessage",
        "module": "android.os.Handler",
        "lineno": 95,
        "index": 6
    }, {
        "filename": "Handler.java",
        "function": "handleCallback",
        "module": "android.os.Handler",
        "lineno": 739,
        "index": 7
    }, {
        "filename": "View.java",
        "function": "run",
        "module": "android.view.View$PerformClick",
        "lineno": 19866,
        "index": 8
    }, {
        "filename": "View.java",
        "function": "performClick",
        "module": "android.view.View",
        "lineno": 4780,
        "index": 9
    }, {
        "filename": "ActionMenuItemView.java",
        "function": "onClick",
        "module": "androidx.appcompat.view.menu.ActionMenuItemView",
        "lineno": 7,
        "index": 10
    }, {
        "filename": "ActionMenuView.java",
        "function": "invokeItem",
        "module": "androidx.appcompat.widget.ActionMenuView",
        "lineno": 4,
        "index": 11
    }, {
        "filename": "MenuBuilder.java",
        "function": "performItemAction",
        "module": "androidx.appcompat.view.menu.MenuBuilder",
        "lineno": 1,
        "index": 12
    }, {
        "filename": "MenuBuilder.java",
        "function": "performItemAction",
        "module": "androidx.appcompat.view.menu.MenuBuilder",
        "lineno": 4,
        "index": 13
    }, {
        "filename": "MenuItemImpl.java",
        "function": "invoke",
        "module": "androidx.appcompat.view.menu.MenuItemImpl",
        "lineno": 15,
        "index": 14
    }, {
        "filename": "MenuBuilder.java",
        "function": "dispatchMenuItemSelected",
        "module": "androidx.appcompat.view.menu.MenuBuilder",
        "lineno": 5,
        "index": 15
    }, {
        "filename": "ActionMenuView.java",
        "function": "onMenuItemSelected",
        "module": "androidx.appcompat.widget.ActionMenuView$MenuBuilderCallback",
        "lineno": 7,
        "index": 16
    }, {
        "filename": "Toolbar.java",
        "function": "onMenuItemClick",
        "module": "androidx.appcompat.widget.Toolbar$1",
        "lineno": 7,
        "index": 17
    }, {
        "filename": "R8$$SyntheticClass",
        "function": "onMenuItemClick",
        "module": "io.sentry.samples.instrumentation.ui.g",
        "lineno": 40,
        "in_app": true,
        "index": 18
    }]"#;

    let modules = format!(
        r#"[{{
        "type": "source",
        "uuid": "{source_id1}"
    }}, {{
        "type": "source",
        "uuid": "{source_id2}"
    }}, {{
        "type": "proguard",
        "uuid": "{proguard_id}"
    }}, {{
        "type": "proguard",
        "uuid": "{missing_proguard_id}"
    }}]"#
    );

    let request = make_jvm_request(
        source,
        r#"{"type": "RuntimeException", "module": "java.lang"}"#,
        frames,
        &modules,
        None,
    );

    let response = symbolication.symbolicate_jvm(request).await;

    assert_snapshot!(response);
}
