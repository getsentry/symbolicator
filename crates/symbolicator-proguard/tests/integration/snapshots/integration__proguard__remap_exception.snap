---
source: crates/symbolicator-proguard/tests/integration/proguard.rs
assertion_line: 122
expression: response
---
exceptions:
  - type: Util$ClassContextSecurityManager
    module: org.slf4j.helpers
stacktraces:
  - frames: []
errors: []
